modules/dag-shared/src/main/scala/org/tessellation/dag/block/BlockValidator.scala

```diff
@@ -0,0 +1,153 @@
+  }
+
+  private def validateParents(signedBlock: Signed[DAGBlock]): F[ValidationResult[Signed[DAGBlock]]] =
+    areParentsAccepted(signedBlock.value).map { result =>
+      val parents = signedBlock.value.parent.toList.toSet
+      if (result.keySet == parents && result.values.forall(identity))
+        signedBlock.validNel[BlockValidationError]
+      else {
+        val notAccepted = result.filterNot { case (_, isAccepted) => isAccepted }.keySet.intersect(parents)
+        BlockParentsNotAccepted(notAccepted).invalidNel
+      }
+    }
+
+  private def validateTransactions(
+    signedBlock: Signed[DAGBlock]
+  ): F[ValidationResult[Signed[DAGBlock]]] =
+    NonEmptyList.fromList(signedBlock.value.transactions.toList) match {
+      case Some(transactions) =>
+        for {
+          propertiesValidation <- validateTransactionsProperties(transactions)
+          transactionChainsValidation <- validateTransactionChains(transactions)
+        } yield propertiesValidation.product(transactionChainsValidation).map(_ => signedBlock)
+      case None => signedBlock.validNel[BlockValidationError].pure[F]
+    }
+
+  private def validateTransactionsProperties(
+    transactions: NonEmptyList[Signed[Transaction]]
+  ): F[ValidationResult[NonEmptyList[Signed[Transaction]]]] =
+    transactionValidator
+      .validate(transactions)
+      .map {
+        case Valid(_)   => transactions.validNel[BlockValidationError]
+        case Invalid(e) => InvalidBlockTransactions(e).invalidNel
+      }
+
+  private def validateTransactionChains(
+    transactions: NonEmptyList[Signed[Transaction]]
+  ): F[ValidationResult[NonEmptyList[Signed[Transaction]]]] =
+    transactions.toList
+      .groupBy(_.value.source)
+      .toSeq
+      .traverse {
+        case (address, txs) =>
+          for {
+            lastAccepted <- getLastAcceptedTransactionRef(address)
+            sorted = txs.sorted
+            chainableTransactions = Consecutive.take(lastAccepted, txs)
+            result = if (sorted == chainableTransactions)
+              txs.validNel[BlockValidationError]
+            else
+              TransactionChainIncorrect(address, NonEmptyList(lastAccepted, sorted.map(_.value.parent))).invalidNel
+          } yield result
+      }
+      .map(_.combineAll.map(_ => transactions))
+}
+
+object BlockValidator {
+  type ValidationResult[A] = ValidatedNel[BlockValidationError, A]
+
+  sealed trait BlockValidationError extends NoStackTrace {
+    val errorMessage: String
+    override def getMessage: String = errorMessage
+  }
+
+  case object InvalidBlockSignatures extends BlockValidationError {
+    override val errorMessage: String = "Block has invalid signatures!"
+  }
+
+  case class InvalidBlockSignaturesCount(is: Int, shouldBe: Int) extends BlockValidationError {
+    override val errorMessage: String = s"Block has invalid signatures number: $is not equal $shouldBe!"
+  }
```

Comment:
maybe ```$is less or equal than $shouldBe```?

Reply:
Right it makes sense if we leave the greater or equal condition in the validator.